### ISSUE TITLE
sharing accessibility extensions (menu, plugin tool, video player)

### ISSUE THREAD

**Issue by armenos** [joomla/accessibility#17]
# sharing accessibility extensions (menu, plugin tool, video player)

Dear JAT’ers 😊
Our dear (beloved) Yair has a question and an proposal on which we need to discuss or just give an answer.
So I did some research and have my answers to Yair about this topic. 
But first I like to share this with you so you can think and discuss about this.

@ylahav said that he has some accessibility extensions which he want to share within Joomla 
Question is, can this be part of Joomla 3 core or not?
Can it be maintained within Joomla?

@ylahav can add if I forgot something to mention. 

The accessibility extensions that Yair wants to give to the community are:
1.	accessibility plugin (today baccessability)
a.	put a float toolbar on the left / right side of the screen
b.	the toolbar give the ability to:
i.	increase / decrease font size
ii.	dark / light contrast
iii.	underlines under all links
STATUS: this plugin is a conversion of the module I wrote and can be found in GITHUB. It was rewritten in order to simply the definition and configuration and to have it as a pure JavaScript (no jquery)
2.	accessibility menu plugin (today bmenu)	
a.	add 'aria' attribute to menu
b.	allow keyboard navigation
c.	pure JS
    STATUS: pass @zwiastunsw Stephan test. can be used.
    Needs: some refinement to simply configuration
3.	accessibility audio / video player (based on JS player which is already accessible) (today bplayer)
a.	has all needed features: subtitle, subscript etc. + keyboard navigation
    STATUS: tested by @zwiastunsw Stefan

**Comment by armenos** [comment]
1.	accessibility plugin (today baccessability)
You can do a pull request to give your code to Joomla Project. Joomla 3.x or Joomla 4.x? 
The next step is to do research about what already exist and see how we can make our own tools or plugins. 
Our documentation for custom elements is good starting point: https://github.com/joomla/accessibility/tree/master/docs/custom-elements
2.	accessibility menu plugin (today bmenu)
It would be better to change the core of Joomla 3.x backend and fronend mod_menu instead of a fixing plugin. You can add the right code in Joomla Project and it will always work fine when people will use Joomla menu module, no fix needed anymore.
You are the HERO 😊	
3.	accessibility audio / video player (based on JS player which is already accessible) (today bplayer)
I don’t know how it exactly works but Joomla Project (Production) has chosen for MediaElement.js You can make your own code and share it on JED, like I said I don’t know how it works.
I think we can also join the new media manager team to support them on a11y questions.
For support on creating pull request or adding code to Joomla, I can’t help with. But I think here are people how can help you to do so.

I'm tagging you guys because you can explain how it works and which pat is the best to take 
@mbabker @zwiastunsw @wilsonge @chmst @tarotray @uglyeoin @ylahav @moje-s @seenee
@Joomv @jonrz @Sandra97 @ciar4n @dgt41 @brianteeman @coolcat-creations @C-Lodder @yvesh

**Comment by C-Lodder** [comment]
> 2. accessibility menu plugin (today bmenu)

This doesn't need to be a plugin, it should just be the default.

> 3. accessibility audio / video player (based on JS player which is already accessible) (today bplayer)

For the video player, we were looking at this as the default: https://github.com/google/shaka-player
I'm not sure how accessible it is, so perhaps give it a test.

**Comment by uglyeoin** [comment]
> It would be better to change the core of Joomla 3.x backend and fronend mod_menu instead of a fixing plugin
Some templates override the menu output in this case the plugin would be preferable.  There are also menu components I think.  I don't know what the official stance is but maybe worth discussing.  I assume we only have to cater for the core.

**Comment by dgrammatiko** [comment]
> I assume we only have to cater for the core.

That's right. We shouldn't try to manipulate in run time other devs code (never, ever)

**Comment by dgrammatiko** [comment]
@C-Lodder is the proposal for shaka-player still valid? I thought it was burned by @mbabker and @wilsonge

**Comment by C-Lodder** [comment]
@dgt41 - I can't remember it being proposed at all.

**Comment by brianteeman** [comment]
1. when i reviewed the original code about six months ago it was not up to the required quality for inclusion in the core AND I am not convinced that this type of thing should be included anyway especially as it is usually considered best to control things like font size in the browser than on the site

2. the core module should be updated if needed. it should not be a plugin. And as @dgt41 says we should never do anything to alter users code.

3. I am assuming this is a front end player - not convinced this should be in core - sounds more like an extension to me - but would need to see the code

**Comment by brianteeman** [comment]
( shaka player is not gpl2 compatible so cannot be used)

**Comment by C-Lodder** [comment]
@brianteeman - It would simply be a replacement for mediaelement (currently being used). It's under the Apache license, so we can use it

**Comment by brianteeman** [comment]
no we cant the apache license is NOT compatible with gpl2 only gpl3

**Comment by C-Lodder** [comment]
ok then, so perhaps update to v3?

**Comment by mbabker** [comment]
We used Bootstrap 2 with its Apache license.  Using precedent, an Apache licensed UI element should be available for consideration, but as with anything licensing related, don't take my word as the last one and ask a "real" expert.

**Comment by brianteeman** [comment]
The "expert" would probably say the media player is code not just ui IANAL

**Comment by C-Lodder** [comment]
Well others were complaining about it: https://github.com/twbs/bootstrap/issues/2054

So Joomla 3 pretty much shouldn't be allowed to be released.

**Comment by ylahav** [comment]
A very nice discussion BUT not the real one.
I will, first, explain the "events":
After having the 'bmenu' plugin which convert any menu to accessible one using pure javascript
and creating a 'bplayer' for accessibility audio / video player, @zwiastunsw asked me what I am going to do with these extensions... and we thought that it could be an idea to have some accessibility extensions, not as part of the core! So, I discuss it with @armenos....
So, first off all the discussion should be fundamental: Does the accessibility group can / it sis a good idea / what ever ... add in its repository extensions for the benefit of the community.
And, yes, this means that we have to think about the quality of the code, copyright etc.
After getting an answer for that we can discuss each specific extension.
Regarding the specific extensions:
1. 'bmenu' - I developed it  for current Joomla version to allow existing Joomla site to make their menu accessible.
2. 'baccessibilty' is a new plugin (based on for my old module functionalities) in pure javascript - the goal is to handle: font sizes, color contrast, underline links, and displaying accessibility statement. Just to late you know that the module is installed in hundreds  of sites in Israel.
3. 'bplayer' - @zwiastunsw asked me if I can develop such a module /plugin - so I just developed it and he tested it.

I also have a (management) component witch allow the user to handle all the content images in one place and maintain the image 'title' and 'alt' values... I am giving it as an example that will help us to help our users making their sites more accessible. 
So, again, first of all, lets focus on the question: is it a good idea that the Joomla community will publish as part of its repository some extensions to help with accessibility.
If NO - the discussion will stop here...
If YES - than we can first answer the questions: where and how - and only after that what...

I hope that my English was clear enough :-) and, please, don't think / write what it is not allowed or not good - think positive - how things can be improved!

**Comment by C-Lodder** [comment]
@ylahav so your bmenu plugin suports ANY markup? Kind of hard to believe it can make any menu, even badly made ones, accessible...

**Comment by brianteeman** [comment]
@ylahav can we see the code - i checked your github repo and it just had the old broken mod_baccessibility - i couldnt find the others

**Comment by ylahav** [comment]
@C-Lodder and @brianteeman - lets focus on the question: does it a good idea (or not) to add a repository with accessibility extensions to Joomla project?
I will be happy to answer or check any specific extension after getting answer to this question.

**Comment by mbabker** [comment]
Unless the project is going to treat those as "official" extensions (same as weblinks or install from web), no, the project should not take on the additional maintenance responsibility of a set of optional extensions for accessibility improvements (even if it might actually benefit users).  This is fine to be left to the extension ecosystem if we can't effectively manage these.

**Comment by dgrammatiko** [comment]
@ylahav back in the day we discussed about the backend menu I gave you the link of the work I'had done at that point https://github.com/joomla/joomla-cms/compare/4.0-dev...dgt41:§4.0-dev-sidebar-menu?expand=1
and I asked you to take ownership and finalise it. Instead of that you created a plugin which is totally out of scope of Joomla core (we care for the code that the project ships, we should not try to patch any other developers code).

Anyways couple questions:
- does the plugin work with all the menu modules in the JED? (by looking at the code I 'll say it's gonna break most of them)
- can we join forces instead of working against each other? Please take ownership of my work above and deliver an accessible menu module?

**Comment by ylahav** [comment]
thanks @mbabker!
I totally agree with you!
From my opinion the 'weblinks' could not be an extra extension if it was not part of the core in first place.
BUT we still have to think on criteria - when the community want / need to support (extra) extension and what will be the rule to maintain such an extension. 

@dgt41 - I might be that I miss understood you - I understood that you are working on it and I didn't want to duplicate the work,.. and yes, we can join force - I will take it and work on it. I will contact you directly.

**Comment by uglyeoin** [comment]
@ylahav I think the issue with a plugin seems to be that it could potentially break or interfere with other 3rd party tools which is why it is better to be part of the core code and not as a separate plugin.  Perhaps I misunderstood the comments.

**Comment by ylahav** [comment]
@uglyeoin - you are forcing something that I tried to avoid in this discussion BUT...
it will not break any 3rd tools because it is activated only on menus (modules) that you want to...
In the configuration you define a list of module that the plugin will work on....
and again, that not the issue, in my opinion in this thread.

**Comment by zwiastunsw** [comment]
1. There is an Accessibility category in JED. The extensions in question can simply be made available on JED in this category. This is not a bad solution. The author takes responsibility for the publication and does not concern us. This is one possibility.
2. We would like to emphasize, however, that these extensions are the result of the work of the Joomla Accessibility Team (and Joomla Project) as a support for addressing accessibility issues that can no longer be fixed by changing the Joomla code. The Menu plugin is a good example. Another is the pagenavigation bar.
3. The question is: **Do these solutions publish as JAT works and how to do it?** 
In my opinion, yes. 
 I suggest that you share them in the [Joomla! Projects repository](https://github.com/joomla-projects), such as custom-elements or com_localise.
